fix XSS vulnerability in scaladoc search by escaping the input parameter

mpollmeier opened the pull request:
Fixes https://github.com/scala/bug/issues/11513

To trigger the XSS vulnerability, simply paste this into the search bar, e.g. on https://www.scala-lang.org/api/2.12.8/:
```
"\><img/src='1'onerror=alert(777111)>{{7*7}}
```

All credit for finding the vulnerability goes to Yeasir Arafat:
skylinearafat@gmail.com
https://www.facebook.com/skylinearafat.arafat

n.b. I've targeted 2.12.x because according to https://github.com/scala/scala/blob/2.13.x/CONTRIBUTING.md this will be merged into 2.13.x 'in time'.

mpollmeier committed:
fix XSS vulnerability in scaladoc search

to trigger XSS vuln, simply paste this into the search bar:
```
"\><img/src='1'onerror=alert(777111)>{{7*7}}
```

all credit for finding the vulnerability goes to *Yeasir Arafat* <skylinearafat@gmail.com>

hrhino commented:
I think they mean "in due time", not "in time for" anything particular.

This isn't really XSS, is it? If you paste some sketchy code into your own browser, that's on you. If there were a way to get a specific scaladoc page to then serve that to other people, that would be a problem.

It's better to escape now, though, while we still can. :+1:

mpollmeier commented:
> If you paste some sketchy code into your own browser, that's on you. 

This is not about the input field in the html, but about how the input parameter is handled. An attacker could send the victim a link like `https://example.com/api/?search=%22\%3E%3Cimg/src=%271%27onerror=alert(777111)%3E{{7*7}}` where the payload instructs the browser to extract userdata for that particular domain (example.com). This is not really critical for https://www.scala-lang.org/api, but for users who host their api docs under their company url.

hrhino commented:
Sorry, I edited instead of posting a new comment. I'm not arguing against this change; I'm simply pointing out that it doesn't appear to be an actual vulnerability affecting scaladoc sites out there right now, unless there's a different attack vector I hadn't thought of.

hrhino commented:
Also, I hate to be a nitpicker, but now I can't search for `:+`.

hrhino commented:
And to prove I'm not just a naysayer, may I suggest:
```diff
diff --git a/src/scaladoc/scala/tools/nsc/doc/html/resource/lib/index.js b/src/scaladoc/scala/tools/nsc/doc/html/resource/lib/index.js
index e899f06b5c..84004ef468 100644
--- a/src/scaladoc/scala/tools/nsc/doc/html/resource/lib/index.js
+++ b/src/scaladoc/scala/tools/nsc/doc/html/resource/lib/index.js
@@ -532,7 +532,6 @@ function searchAll() {
     scheduler.clear("search"); // clear previous search
     maxJobs = 1; // clear previous max
     var searchStr = $("#textfilter input").attr("value").trim() || '';
-    searchStr = escape(searchStr);
 
     if (searchStr === '') {
         $("div#search-results").hide();
@@ -562,10 +561,11 @@ function searchAll() {
     entityH1.innerHTML = "Entity results";
     entityResults.appendChild(entityH1);
 
-    $("div#results-content")
-        .prepend("<span class='search-text'>"
-                +"  Showing results for <span class='query-str'>\"" + searchStr + "\"</span>"
-                +"</span>");
+    var searchTextEl = $("<span class='search-text'>"
+                        +"  Showing results for <span class='query-str' />"
+                        +"</span>");
+    searchTextEl.find(".query-str").text(searchStr);
+    $("div#results-content").prepend(searchTextEl);
```

NthPortal commented:
I would personally be inclined to never create the tag with any contents using jQuery (or anything else, for that matter) - always setting the text outside of the tag means you will never be able to accidentally open yourself up to injection.

In this case I would do one of the following (even though it's slightly more verbose):

```javascript
$("div#results-content")
    .prepend(
        $("<span>", {class: "search-text"})
            .append(document.createTextNode("  Showing results for "))
            .append($("<span>", {class: "query-str", text: searchStr}))
    );
```

```javascript
$("div#results-content")
    .prepend(
        $("<span>", {class: "search-text"})
            .append(document.createTextNode("  Showing results for "))
            .append($("<span>", {class: "query-str"}).text(searchStr))
    );
```

julienrf commented on src/scaladoc/scala/tools/nsc/doc/html/resource/lib/index.js:
```diff
@@ -532,6 +532,7 @@ function searchAll() {
     scheduler.clear("search"); // clear previous search
     maxJobs = 1; // clear previous max
     var searchStr = $("#textfilter input").attr("value").trim() || '';
+    searchStr = escape(searchStr);
```
What about using `encodeURIComponent` instead? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/escape

NthPortal replied:
I don't think this is the correct solution regardless - as pointed about by Harrison, it prevents you from searching for some things (like `:+`). The problem is that the untrusted string is embedded in a tag creation below.

jdelta-RBS commented:
@mpollmeier I tried this in Chrome, FireFox, IE -> nothing. It *does* work as originally posted, but it is not a realistic attack vector that would warrant any type of change.

mpollmeier committed:
fix xss by writing the input parameter properly to the dom

rather than escaping the search string, which breaks the search for
e.g. `:+`

solution contributed by NthPortal in https://github.com/scala/scala/pull/8018#issuecomment-488546695

hrhino commented on src/scaladoc/scala/tools/nsc/doc/html/resource/lib/index.js:
```diff
@@ -562,9 +562,12 @@ function searchAll() {
+          $("<span>")
+              .addClass("search-text")
+              .append(document.createTextNode("  Showing results for "))
+              .append($("<span>").addClass("query-str").text(searchStr))
```
So that's how it's done. Mine felt like a hack, but it's been so long since I've used jQuery (Ext.js for life!) that I forgot the right way.

NthPortal replied:
I heavily abused google and the browser console to figure it out

mpollmeier commented:
Thanks for all the feedback! I updated the PR to use @NthPortal's 2nd solution from https://github.com/scala/scala/pull/8018#issuecomment-488546695, which IMO is the cleanest approach.

mpollmeier committed:
four space indentation

hrhino reviewed (approved):
Published the doc site locally; all appears well. I can search for my silly symbolic aliases, and I can't give myself popups, so LGTM.

Thanks!

hrhino commented:
/synch

lrytz commented:
Thanks, everyone involved!
